pallets/services/src/lib.rs

```diff
@@ -0,0 +1,721 @@
+			/// The ID of the service.
+			service_id: u64,
+			/// The ID of the service blueprint.
+			blueprint_id: u64,
+		},
+
+		/// A job has been called.
+		JobCalled {
+			/// The account that called the job.
+			caller: T::AccountId,
+			/// The ID of the service.
+			service_id: u64,
+			/// The ID of the call.
+			call_id: u64,
+			/// The index of the job.
+			job: u8,
+			/// The arguments of the job.
+			args: Vec<Field<T::AccountId>>,
+		},
+
+		/// A job result has been submitted.
+		JobResultSubmitted {
+			/// The account that submitted the job result.
+			provider: T::AccountId,
+			/// The ID of the service.
+			service_id: u64,
+			/// The ID of the call.
+			call_id: u64,
+			/// The index of the job.
+			job: u8,
+			/// The result of the job.
+			result: Vec<Field<T::AccountId>>,
+		},
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	// Counters
+
+	/// The next free ID for a service blueprint.
+	#[pallet::storage]
+	#[pallet::getter(fn next_blueprint_id)]
+	pub type NextBlueprintId<T> = StorageValue<_, u64, ValueQuery>;
+
+	/// The next free ID for a service request.
+	#[pallet::storage]
+	#[pallet::getter(fn next_service_request_id)]
+	pub type NextServiceRequestId<T> = StorageValue<_, u64, ValueQuery>;
+
+	/// The next free ID for a service Instance.
+	#[pallet::storage]
+	#[pallet::getter(fn next_instance_id)]
+	pub type NextInstanceId<T> = StorageValue<_, u64, ValueQuery>;
+
+	/// The next free ID for a service call.
+	#[pallet::storage]
+	#[pallet::getter(fn next_job_call_id)]
+	pub type NextJobCallId<T> = StorageValue<_, u64, ValueQuery>;
+
+	/// The service blueprints along with their owner.
+	#[pallet::storage]
+	#[pallet::getter(fn blueprints)]
+	pub type Blueprints<T: Config> = StorageMap<
+		_,
+		Identity,
+		u64,
+		(T::AccountId, ServiceBlueprint),
+		ResultQuery<Error<T>::BlueprintNotFound>,
+	>;
+
+	/// The service providers for a specific service blueprint.
+	#[pallet::storage]
+	#[pallet::getter(fn service_providers)]
+	pub type ServiceProviders<T: Config> = StorageDoubleMap<
+		_,
+		Identity,
+		u64,
+		Identity,
+		T::AccountId,
+		ServiceProviderPrefrences,
+		ResultQuery<Error<T>::NotRegistered>,
+	>;
+
+	/// The service requests along with their owner.
+	/// Request ID -> Service Request
+	#[pallet::storage]
+	#[pallet::getter(fn service_requests)]
+	pub type ServiceRequests<T: Config> = StorageMap<
+		_,
+		Identity,
+		u64,
+		ServiceRequest<T::AccountId, BlockNumberFor<T>>,
+		ResultQuery<Error<T>::ServiceRequestNotFound>,
+	>;
+
+	/// The Services Instances
+	/// Service ID -> Service
+	#[pallet::storage]
+	#[pallet::getter(fn services)]
+	pub type Instances<T: Config> = StorageMap<
+		_,
+		Identity,
+		u64,
+		Service<T::AccountId, BlockNumberFor<T>>,
+		ResultQuery<Error<T>::ServiceNotFound>,
+	>;
+
+	/// The Service Job Calls
+	/// Service ID -> Call ID -> Job Call
+	#[pallet::storage]
+	#[pallet::getter(fn job_calls)]
+	pub type JobCalls<T: Config> = StorageDoubleMap<
```

Comment:
Do we want to provide a potential way for historical job calls in the pallet? Possibly if the user requests a service where they persist all the historical data?

Reply:
Agreed, that should be added as a TTL for job calls.